clrtrust-helper.c

```diff
@@ -0,0 +1,148 @@
+    puts(help);
+}
+
+int main(int argc, char **argv) {
+
+    runmode_t runmode = MODE_INVALID; /* global switch: whether to produce fingerprint
+                                   or subject hash. */
+
+    BIO *inbio = NULL;
+
+    /* fingerprint-related variables */
+    const EVP_MD *finger_type = NULL;
+    unsigned int finger_sz;
+    unsigned char finger[EVP_MAX_MD_SIZE];
+
+    /* subjecthash-related variables */
+    unsigned long subject_hash;
+
+    unsigned int i;
+    char c;
+
+    char *fname = NULL;
+    size_t sz = 0;
+
+    while ((c = getopt(argc, argv, "fs")) != -1) {
+        switch (c) {
+            case 'f':
+                if (runmode == MODE_INVALID) {
+                    runmode = MODE_FINGER;
+                } else {
+                    print_help();
+                    return 1;
+                }
+                break;
+            case 's':
+                if (runmode == MODE_INVALID) {
+                    runmode = MODE_HASH;
+                } else {
+                    print_help();
+                    return 1;
+                }
+                break;
+            default:
+                printf("Unrecognized option -%c.\n", c);
+                print_help();
+                return 1;
+        }
+    }
+    
+    if (runmode == MODE_INVALID) {
+        print_help();
+        return 1;
+    }
+
+    OpenSSL_add_all_algorithms();
+
+    finger_type = EVP_sha1();
+    
+    inbio = BIO_new(BIO_s_file());
+
+    while (getline(&fname, &sz, stdin) != -1) {
+        int fname_len = strlen(fname);
+        int err = 1;
+        FILE *fp = NULL;
+        X509 *cert = NULL;
+
+        if (fname_len < 1) continue;
+        if (fname[fname_len-1] == '\n') {
+            fname[fname_len-1] = '\0';
+        }
+
+        if (!(fp = fopen(fname, "r"))) {
+            goto wrap_up;
+        }
+
+        if (!BIO_set_fp(inbio, fp, BIO_NOCLOSE)) {
+            goto wrap_up;
+        }
+
+        if (!(cert = PEM_read_bio_X509(inbio, NULL, 0, NULL))) {
+            goto wrap_up;
+        }
+
+        switch (runmode) {
+            case MODE_FINGER:
+                if (!X509_digest(cert, finger_type, finger, &finger_sz)) {
+                    goto wrap_up;
+                }
+                printf("%s\t", fname);
+                for (i=0; i < finger_sz; i++) {
+                    printf("%02X", finger[i]);
+                    if (i < finger_sz-1) printf(":");
+                }
+                putc('\n', stdout);
+                err = 0;
+                break;
+            case MODE_HASH:
+                subject_hash = X509_subject_name_hash(cert);
+                printf("%s\t%08lx\n", fname, subject_hash);
+                err = 0;
+                break;
+            default:
+                return 1;
+        }
+
+wrap_up:
+        if (fp) fclose(fp);
+        if (cert) X509_free(cert);
+        if (err) printf("%s\tERROR\n", fname);
+    }
+
+    if (fname) free(fname);
+
+    return 0;
```

Comment:
Perhaps a global return code to indicate if there were any errors (e.g. missing files) rather than success except for "somehow I got to line 136" and "missing or conflicting -s -f switches"?